### ISSUE TITLE
Support `Import from CSV` option

### ISSUE THREAD

**Issue by apexdodge** [RaythaHQ/raytha#81]
#79 is a pre-req

Add a new item in this dropdown box: `Import from CSV`

![image](https://user-images.githubusercontent.com/777005/221736541-0c5f95f6-e755-41d4-9e8f-2f318681c619.png)

The header column in the CSV should match the DeveloperName of the field it should map to.

There should be two options:

- Insert new records only
- Insert and update existing records

**Comment by smd-11** [comment]
Hey @apexdodge, I would like to work on this.

**Comment by smd-11** [comment]
I am good

**Comment by apexdodge** [comment]
@smd-11 Okay great. We can split the task up into 2 parts. We can start with part 1 which I think is the harder of the two actually, the backend.

1) Create a new Command called `BeginImportContentItemsFromCsv` https://github.com/RaythaHQ/raytha/tree/main/src/Raytha.Application/ContentItems/Commands. Can use the BeginExportContentItemsToCsv command as a starting place.

2) The Command should probably take in the following attributes:

- ShortGuid ViewId
- string ImportMethod ("insert_new_only", "upsert_all_records", "update_existing_only")
- bool ImportAsDraft
- byte[] CsvAsBytes

3) The Handler can actually look a lot like ExportToCsv's handler.

4) BackgroundTask for this should be a lot simpler ExportToCsv's. Most of the logic there can be written from scratch but follow this pattern:

Challenges in this step

- Converting the CsvAsBytes to an ienumerable<ienumerable<string>> so that the data is iterable.
- Get the first row which should be columns of `id` and the rest of the columns should match a Developer Name as defined by the user in a content type.
- Loop through iterable
    - depending on ImportMethod:
    - if value exists in the `id` column, get the item, update item
    - if value does not exist in `id` column, insert item
    - if ImportAsDraft, set entity.IsDraft = true and entity.IsPublished = false. Else, entity.IsDraft = false and entity.IsPublished = true

Some advice, up to you:

- Review docs on background tasks in Raytha: https://docs.raytha.com/articles/background_tasks.html
- Look at CreateContentItem and EditContentItem commands:
https://github.com/RaythaHQ/raytha/blob/main/src/Raytha.Application/ContentItems/Commands/CreateContentItem.cs
https://github.com/RaythaHQ/raytha/blob/main/src/Raytha.Application/ContentItems/Commands/EditContentItem.cs
- In step 4, at first, do not try to convert bytes to CSV and do any loop through. Just hardcode inserting values, just to see if you can get items to insert at all
- Feel free to create a temporary endpoint on a controller to invoke the command without any UI.

Thank you again for your interest and let me know if you have any questions or concerns.

**Comment by smd-11** [comment]
Hey @apexdodge ,getting exception at line 49 in QueueHostedService.cs :

The requested operation requires an element of type 'Number', but the target element has type 'String'.

Can you help with this?

![image](https://github.com/RaythaHQ/raytha/assets/32430772/3a5b5230-8f6f-495e-abd8-41fe721754cb)

**Comment by apexdodge** [comment]
@smd-11 This code is responsible for seeing if there were any requests to process a background task, and if there is, it will begin that execution.

I suspect there might be an issue with the BackgroundTask class you may have created as shown here: https://docs.raytha.com/articles/background_tasks.html#create-a-class-that-implements-the-iexecutebackgroundtask-interface -- perhaps a mismatch in pulling the arguments out.

Can you share your BeginImportContentItemsFromCsv file to this point? I think it might help uncover the issue.

**Comment by smd-11** [comment]
resolved issue.... was using byte instead of byte[]

**Comment by smd-11** [comment]
@apexdodge are these fields enough to be there in csv file. 

**Id,Template,title,content,contenttype**
Test Data,Content item detail view,Test Data,This is test data,pages

**Comment by apexdodge** [comment]
@smd-11 

Good questions 👍🏻 .

contenttype column is not needed because when you click `Import from CSV`, we are doing it from a View that is already scoped to the content type (hence the ViewId, which we can use to get the contenttype): 

![image](https://github.com/RaythaHQ/raytha/assets/777005/864c6871-98f7-4618-91f0-2b51fb406c3c)

The rest are good.

One change on this:

![image](https://github.com/RaythaHQ/raytha/assets/777005/dd646670-6cf4-4546-94ed-2421c0a4988a)

I think this should be the Template's developer name, not label. You can find the developer name of a template like so: 

![image](https://github.com/RaythaHQ/raytha/assets/777005/089185a2-a572-4706-8329-127da9dc52ad)

Then you can use something like below (doing it from memory)

```
var template = _db.WebTemplates.FirstOrDefault(p => p.DeveloperName == templateDeveloperName)
```

Hope that helps! Let me know if you have any other questions.

**Comment by smd-11** [comment]
Hey @apexdodge, pushed code to https://github.com/smd-11/raytha/tree/excfix.

**Comment by apexdodge** [comment]
@smd-11 

Excellent work. There is one place that needs fixing.

![image](https://github.com/RaythaHQ/raytha/assets/777005/a9997b50-97a0-44c3-9af9-184870a3f92d)

We cannot hard code these since users define their own content types and create their own fields. So we are expecting the columns in the CSV to match developer names as shown here:

![image](https://github.com/RaythaHQ/raytha/assets/777005/44c90776-87cc-4e64-92f1-09bec31a141c)

So you can create your own fields first in the user interface, specify the developer name, and then use CSV import with column names that match those developer names so that you can bulk import data.

Let me know if that makes sense.

**Comment by smd-11** [comment]
yeah will update code accordingly

**Comment by smd-11** [comment]
Hi @apexdodge ,

updated code at [(https://github.com/smd-11/raytha/tree/excfix)]..Have a look and let me know if any changes required

**Comment by smd-11** [comment]
@apexdodge 

A few queries:

-  should data for **attachment** field type be uploaded from csv? If yes, what the user will enter in csv "Local path of file"?
-  can we have only text data in csv for **wysiwyg and long_text** field type or images too? if image too then how can it will be entered in csv?

**Comment by apexdodge** [comment]
@smd-11, yes good questions.

- For v1 of this feature we will not support file upload. I have ideas on how it will work, but it's too much for right now.
- Only text data is fine, or more specifically, the expectation is that the person has html in these fields. So don't worry about images here.
- For one to one relationship, it is expecting the Id of the related object item. Similar to how we have the Id field for Updates, this column should point to the Id of another content item.

Let me know if anything isn't clear from that, I can happily explain.

**Comment by smd-11** [comment]
@apexdodge  created PR for this.

**Comment by apexdodge** [comment]
@smd-11 Excellent, thank you. I will review this weekend.

**Comment by apexdodge** [comment]
Branch `[release-v1.1.2](https://github.com/RaythaHQ/raytha/tree/release-v1.1.2)` has been published and I did a little bit of code clean up and testing. It works quite well.

Tomorrow I will merge into `main` and write up the documentation.

**Comment by apexdodge** [comment]
Released with v1.1.2!
